### PR TITLE
Support other LSPs completion icons

### PR DESCRIPTION
**Pull request by badosu**
This is an initial effort to show how support for other LSP plugins may work.

Screenshot from 2019-06-04 14-57-50

This implementation works, but would gladly appreciate feedback if a better approach is possible.

See also:

- https://github.com/vhakulinen/gnvim/issues/47#issuecomment-498773230
- https://github.com/neovim/neovim/issues/10119

### CODE REVIEW

**Comment by vhakulinen** [comment]
I think this is good enough solution. I would like to see the missing ones added too (folder and file for example).

**Comment by badosu** [comment]
@vhakulinen What do you mean, the coc ones? Coc.nvim has less granularity than vim-lsp so many icons are not mapped.

I used the ones I thought made sense, if you want to make a suggestion see https://github.com/neoclide/coc.nvim/blob/909710fddb04d383e5546b0f869c44f395a80d02/src/languages.ts#L143-L167 and let me know which icon you think is mapped better.

**Comment by vhakulinen** [comment]
Hmh, lets get back to this once #50 is merged.

**Comment by badosu** [comment]
@vhakulinen @smolck rebased.

I did not have found the time to test languageclient-neovim, will do that when I can. Meanwhile I hope we can merge this.

**Comment by smolck** [comment]
@badosu I was able to test this with LanguageClient-Neovim, and it appeared to work and look as intended. Here is the diff:
```diff
diff --git a/src/nvim_bridge.rs b/src/nvim_bridge.rs
index be8e98a..6572334 100644
--- a/src/nvim_bridge.rs
+++ b/src/nvim_bridge.rs
@@ -244,31 +244,31 @@ pub enum CompletionItemKind {
 impl From<&str> for CompletionItemKind {
     fn from(from: &str) -> Self {
         match from {
-            "class" | "C" => CompletionItemKind::Class,
+            "class" | "C" | "Class" => CompletionItemKind::Class,
             "color" => CompletionItemKind::Color,
-            "constant" => CompletionItemKind::Constant,
-            "constructor" => CompletionItemKind::Constructor,
-            "enum" => CompletionItemKind::Enum,
-            "enum member" => CompletionItemKind::EnumMember,
-            "event" | "E" => CompletionItemKind::Event,
-            "function" => CompletionItemKind::Function,
-            "file" | "F" => CompletionItemKind::File,
-            "folder" => CompletionItemKind::Folder,
+            "constant" | "Constant" => CompletionItemKind::Constant,
+            "constructor" | "Constructor" => CompletionItemKind::Constructor,
+            "enum" | "Enum"=> CompletionItemKind::Enum,
+            "enum member" | "Enum Member" => CompletionItemKind::EnumMember,
+            "event" | "E" | "Event"=> CompletionItemKind::Event,
+            "function" | "Function" => CompletionItemKind::Function,
+            "file" | "F" | "File" => CompletionItemKind::File,
+            "folder" | "Folder" => CompletionItemKind::Folder,
             "field" | "m" => CompletionItemKind::Field,
-            "interface" | "I" => CompletionItemKind::Interface,
-            "keyword" | "k" => CompletionItemKind::Keyword,
-            "method" | "f" => CompletionItemKind::Method,
-            "module" | "M" => CompletionItemKind::Module,
-            "operator" | "O" => CompletionItemKind::Operator,
-            "property" => CompletionItemKind::Property,
+            "interface" | "I" | "Interface" => CompletionItemKind::Interface,
+            "keyword" | "k" | "Key" => CompletionItemKind::Keyword,
+            "method" | "f" | "Method" => CompletionItemKind::Method,
+            "module" | "M" | "Module" => CompletionItemKind::Module,
+            "operator" | "O" | "Operator" => CompletionItemKind::Operator,
+            "property" | "Property" => CompletionItemKind::Property,
             "reference" | "r" => CompletionItemKind::Reference,
             "snippet" => CompletionItemKind::Snippet,
-            "struct" | "S" => CompletionItemKind::Struct,
+            "struct" | "S" | "Struct" => CompletionItemKind::Struct,
             "text" => CompletionItemKind::Text,
-            "type parameter" | "T" => CompletionItemKind::TypeParameter,
+            "type parameter" | "T" | "Type Parameter" => CompletionItemKind::TypeParameter,
             "unit" | "U" => CompletionItemKind::Unit,
             "value" => CompletionItemKind::Value,
-            "variable" | "v" => CompletionItemKind::Variable,
+            "variable" | "v" | "Variable" => CompletionItemKind::Variable,
             _ => CompletionItemKind::Unknown,
         }
     }
```
Considering that this is a small addition, I think we could probably just include the support for LanguageClient-Neovim in this PR. What do you think?

**Comment by badosu** [comment]
No problem, I'll include it, thanks!

**Comment by badosu** [comment]
@vhakulinen @smolck I updated to use LanguaceClient-neovim icons as well.

Included some missing kinds from vim-lsp and took the liberty to merge some kinds icons.

Let me know your thoughts.

**Comment by badosu** [comment]
Just pushed a separate commit for ease-of-review and reversibility extracting `CompletionItemKind` from `nvim_bridge`.

**Comment by vhakulinen** [comment]
Interesting that LanguageClient-neovim doesn't comply to the lsp documentation: https://microsoft.github.io/language-server-protocol/specification#textDocument_completion. This PR should not add any new keywords, since the ones documented in the lsp spec are already there.

File: src/ui/popupmenu/completion_item_kind.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#[derive(Clone)]
+pub enum CompletionItemKind {
```
Comment (vhakulinen) [inline src/ui/popupmenu/completion_item_kind.rs]: Keep this in the nvim_bridge file.

Comment (badosu) [reply src/ui/popupmenu/completion_item_kind.rs]: Done

**Comment by badosu** [comment]
> This PR should not add any new keywords, since the ones documented in the lsp spec are already there.

You mean, we should restrict only to kinds specified by the documentation?

```
namespace CompletionItemKind {
	export const Text = 1;
	export const Method = 2;
	export const Function = 3;
	export const Constructor = 4;
	export const Field = 5;
	export const Variable = 6;
	export const Class = 7;
	export const Interface = 8;
	export const Module = 9;
	export const Property = 10;
	export const Unit = 11;
	export const Value = 12;
	export const Enum = 13;
	export const Keyword = 14;
	export const Snippet = 15;
	export const Color = 16;
	export const File = 17;
	export const Reference = 18;
	export const Folder = 19;
	export const EnumMember = 20;
	export const Constant = 21;
	export const Struct = 22;
	export const Event = 23;
	export const Operator = 24;
	export const TypeParameter = 25;
}
```

**Comment by vhakulinen** [comment]
Yes. Otherwise we'll end up supporting unmaintainable set of completion kinds. We could also write our own documented set of supported kinds or even create a way to set the map from kind value to an icon through viml. But for now, lets stick with the kinds documented in lsp spec.

**Comment by badosu** [comment]
@vhakulinen It seems this is a WIP at the moment, but some clients already have it implemented, see: https://github.com/microsoft/language-server-protocol/issues/343

Anyway, I reverted back to only standard kinds.

**Comment by badosu** [comment]
Let me know if there are any issues pending.

**Comment by smolck** [comment]
LGTM. Will definitely be nice to support the main LSPs 👍

**Comment by vhakulinen** [comment]
Merged!

**Pull request merged**